# Table of Content


1. **[Importing Modules (Pandas)](#pandas)**
<br><br> 
2. **[Pandas DataFrame](#dataframes)**
<br><br>
3. **[Manipulating DataFrame](#dataframes)**
<br><br>
4. **[Reading Data from Different Sources](#reading_data)**



<a id="pandas"> </a>
# 1. Pandas

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> Pandas contain data structures and data manipulation tools designed for data cleaning and analysis.
<br><br>
                       Pandas is designed for working with tabular data.<br><br>

A module/library in python is simply a way to organize the code, and it contains either python classes or just functions. 

### Quick look at functions

In [72]:
# what are functions and arguments
def add(x,y):
    c = x+y
    return c

In [73]:
import cal

In [74]:
#Explain functions and how import works to students
import inspect
print(inspect.getsource(cal.add))

def add(x,y=5):
    c = x+y
    return c



In [75]:
add(2,3)

5

In [76]:
cal.cross(2,3)

6

**How to install and import pandas?**<br>
1. Install pandas:<br><br>
`!pip install pandas`<br><br>
2. Import pandas:<br><br>
`import pandas as pd`

In [77]:
#Check the list of base packages
!pip list

Package                       Version
----------------------------- --------------------
alabaster                     0.7.12
anaconda-client               1.11.0
anaconda-navigator            2.3.2
anaconda-project              0.11.1
anyio                         3.5.0
appdirs                       1.4.4
argon2-cffi                   21.3.0
argon2-cffi-bindings          21.2.0
arrow                         1.2.2
astroid                       2.11.7
astropy                       5.1
atomicwrites                  1.4.0
attrs                         21.4.0
Automat                       20.2.0
autopep8                      1.6.0
Babel                         2.9.1
backcall                      0.2.0
backports.functools-lru-cache 1.6.4
backports.tempfile            1.0
backports.weakref             1.0.post1
bcrypt                        3.2.0
beautifulsoup4                4.11.1
binaryornot                   0.4.4
bitarray                      2.5.1
bkcharts                      0.2
blac

In [78]:
# install pandas (one time only)
# !pip install pandas

In [79]:
#import pandas 'library/package/modul
import pandas as pd

 `as` is used as an alias in pandas. So from now on we will use `pd.` instead of `pandas.` 
 
<br>
<span style="color:crimson">Always use libraries if they are freely available. It saves time, and those codes are already tested, debugged and optimized.</span>

<a id="dataframes"> </a>
# 2. Pandas DataFrames

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> A DataFrame is a tabular representation of data containing an ordered collection of columns, each of which can be a different type (numeric, string, boolean, and so on). <br><br>                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## <span style="color:darkgreen;">To read data from a csv file</span>

In [80]:
# read the example.csv file in a dataframe
df=pd.read_csv('BMI.csv')
df.head()

Age  Weight (in kg)  Height (in m)
0   45              60           1.35
1   12              43           1.21
2   54              78           1.50
3   26              65           1.21
4   68              50           1.32

In [81]:
# check the type
type(df)

pandas.core.frame.DataFrame

On checking the data type, we notice it is read as pandas data frame.

## <span style="color:darkgreen;">To print top & bottom rows of the data</span>

In [82]:
#top 3 rows
df.head(3)

Age  Weight (in kg)  Height (in m)
0   45              60           1.35
1   12              43           1.21
2   54              78           1.50

By default, the `.head()` will display **first** five rows. However, we can set the desired number of rows to be displayed.

In [83]:
#bottom 3 rows
df.tail(3)

Age  Weight (in kg)  Height (in m)
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

By default, the `.tail()` will display **last** five rows. However, we can set the desired number of rows to be displayed.

## <span style="color:darkgreen;">To obtain the dimensions of the data</span>

In [84]:
# get the shape of the dataset
df.shape

(23, 3)

## <span style="color:darkgreen;">To know the data types of a data frame</span>

In [85]:
# get all the data types
df.dtypes

Age                 int64
Weight (in kg)      int64
Height (in m)     float64
dtype: object

We see the data type of each variable.

## <span style="color:darkgreen;">Print more information about the data</span>

In [86]:
# All information together
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             23 non-null     int64  
 1   Weight (in kg)  23 non-null     int64  
 2   Height (in m)   23 non-null     float64
dtypes: float64(1), int64(2)
memory usage: 680.0 bytes


We see this output gives the number of rows present in the data `RangeIndex: 23 entries, 0 to 22` There are 23 rows numbered from 0 to 22. And there are a total of three columns - `Data columns (total 3 columns)`. 

Consider `Age 23 non-null int64` indicates that the column named 'Age' has 23 non-null observations having the data type 'int64'

And finally the memory used to save this dataframe is 680 bytes.

In [87]:
# describe your data
df.describe()

Age  Weight (in kg)  Height (in m)
count  23.000000       23.000000      23.000000
mean   44.521739       58.304348       1.528261
std    20.586557       19.401112       0.227309
min    10.000000       21.000000       1.210000
25%    26.000000       44.000000       1.320000
50%    54.000000       65.000000       1.520000
75%    62.000000       76.000000       1.690000
max    75.000000       89.000000       1.850000

# `.loc` and `.iloc` methods

## <span style="color:darkgreen;">Indexing a dataframe using `.loc`</span>

`DataFrame.loc[]` is label-based method, which means that you have to specify rows and columns based on their row and column labels.

In [88]:
#syntax
dataframe.loc[row-start:row-end, column-start:column-end]

NameError: name 'dataframe' is not defined

In [89]:
# get first row with all columns
df.loc[0]

Age               45.00
Weight (in kg)    60.00
Height (in m)      1.35
Name: 0, dtype: float64

In [90]:
# get the value of Age in the 1st row
df.loc[0,'Age']

45

## <span style="color:darkgreen;">Selecting multiple rows</span>

In [91]:
# select row 4, 7, 10
df.loc[[4,7,10]]

Age  Weight (in kg)  Height (in m)
4    68              50           1.32
7    57              34           1.61
10   23              53           1.50

We use two square brackets since we are passing a list of row numbers to be accessed.

## <span style="color:darkgreen;">Selecting a range of rows</span>

In [92]:
# select rows from 11 to 18
df.loc[10:18]

Age  Weight (in kg)  Height (in m)
10   23              53           1.50
11   34              65           1.76
12   55              89           1.65
13   23              45           1.75
14   56              76           1.69
15   67              78           1.85
16   26              65           1.21
17   56              74           1.69
18   67              78           1.85

## <span style="color:darkgreen;">Selecting the first column</span>

In [93]:
# select 1st column with all rows
df.loc[:,'Age']

0     45
1     12
2     54
3     26
4     68
5     21
6     10
7     57
8     75
9     32
10    23
11    34
12    55
13    23
14    56
15    67
16    26
17    56
18    67
19    26
20    68
21    56
22    67
Name: Age, dtype: int64

To select the last column we use -1, to select the second last column we use -2

## <span style="color:darkgreen;">Select the first two columns</span>

In [94]:
# select the first two columns with all rows
df.loc[:,['Age','Weight (in kg)']]

Age  Weight (in kg)
0    45              60
1    12              43
2    54              78
3    26              65
4    68              50
5    21              43
6    10              32
7    57              34
8    75              23
9    32              21
10   23              53
11   34              65
12   55              89
13   23              45
14   56              76
15   67              78
16   26              65
17   56              74
18   67              78
19   26              65
20   68              50
21   56              76
22   67              78

## <span style="color:darkgreen;">Indexing a dataframe using `.iloc`</span>

`DataFrame.iloc[]` is integer position-based, so you have to specify rows and columns by their integer position values (0-based integer position).

**Note:** the row names are numbers 

In [95]:
# using iloc select 'Age' & 'Weight (in kg)'
df.iloc[:,0:2]

Age  Weight (in kg)
0    45              60
1    12              43
2    54              78
3    26              65
4    68              50
5    21              43
6    10              32
7    57              34
8    75              23
9    32              21
10   23              53
11   34              65
12   55              89
13   23              45
14   56              76
15   67              78
16   26              65
17   56              74
18   67              78
19   26              65
20   68              50
21   56              76
22   67              78

<a id="manipulatingDF"> </a>
# 3. Manipulating a Dataframe

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> CAUTION:<br>
                        1. DataFrame[column] works for any column name, but DataFrame.column only works when the column name is a valid Python variable name.<br>
                        2. New columns cannot be created with the ` data.BMI ` syntax.
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## <span style="color:darkgreen;">Adding a new column to the dataframe</span>

columns/variables/features mean the same

In [96]:
# create a new column BMI which is given by weight / H**2
df['BMI']=df["Weight (in kg)"]/df['Height (in m)']**2
df

Age  Weight (in kg)  Height (in m)        BMI
0    45              60           1.35  32.921811
1    12              43           1.21  29.369579
2    54              78           1.50  34.666667
3    26              65           1.21  44.395875
4    68              50           1.32  28.696051
5    21              43           1.52  18.611496
6    10              32           1.65  11.753903
7    57              34           1.61  13.116778
8    75              23           1.24  14.958377
9    32              21           1.52   9.089335
10   23              53           1.50  23.555556
11   34              65           1.76  20.983988
12   55              89           1.65  32.690542
13   23              45           1.75  14.693878
14   56              76           1.69  26.609713
15   67              78           1.85  22.790358
16   26              65           1.21  44.395875
17   56              74           1.69  25.909457
18   67              78           1.85  22.790358
19   26              65           1.21  44.395875
20   68              50           1.32  28.696051
21   56              76           1.69  26.609713
22   67              78           1.85  22.790358

In [97]:
# check the shape of the data
df.shape

(23, 4)

## <span style="color:darkgreen;">Adding a new row to the dataframe</span>

In [98]:
# add a new row to the data with values [56, 76, 1.69, 26.609713] using .loc method
df.loc[23]=[56,76,1.69,26.609713]

In [99]:
df.shape

(24, 4)

In [100]:
df['BMI']=round(df['BMI'],2)

In [101]:
df

Age  Weight (in kg)  Height (in m)    BMI
0   45.0            60.0           1.35  32.92
1   12.0            43.0           1.21  29.37
2   54.0            78.0           1.50  34.67
3   26.0            65.0           1.21  44.40
4   68.0            50.0           1.32  28.70
5   21.0            43.0           1.52  18.61
6   10.0            32.0           1.65  11.75
7   57.0            34.0           1.61  13.12
8   75.0            23.0           1.24  14.96
9   32.0            21.0           1.52   9.09
10  23.0            53.0           1.50  23.56
11  34.0            65.0           1.76  20.98
12  55.0            89.0           1.65  32.69
13  23.0            45.0           1.75  14.69
14  56.0            76.0           1.69  26.61
15  67.0            78.0           1.85  22.79
16  26.0            65.0           1.21  44.40
17  56.0            74.0           1.69  25.91
18  67.0            78.0           1.85  22.79
19  26.0            65.0           1.21  44.40
20  68.0            50.0           1.32  28.70
21  56.0            76.0           1.69  26.61
22  67.0            78.0           1.85  22.79
23  56.0            76.0           1.69  26.61

We see that a new row number 23 has be added to the data.

## <span style="color:darkgreen;">Sorting the dataframe</span>

In [102]:
pd??

In [103]:
# sort the data frame on basis of 'Age' values, by default the values will get sorted in ascending order
# Note: 'ascending = False' will sort the data frame in descending order.
df.sort_values('Age', ascending = True, inplace=True)


## <span style="color:darkgreen;">Droping Rows and Columns</span>

In [104]:
df

Age  Weight (in kg)  Height (in m)    BMI
6   10.0            32.0           1.65  11.75
1   12.0            43.0           1.21  29.37
5   21.0            43.0           1.52  18.61
13  23.0            45.0           1.75  14.69
10  23.0            53.0           1.50  23.56
19  26.0            65.0           1.21  44.40
3   26.0            65.0           1.21  44.40
16  26.0            65.0           1.21  44.40
9   32.0            21.0           1.52   9.09
11  34.0            65.0           1.76  20.98
0   45.0            60.0           1.35  32.92
2   54.0            78.0           1.50  34.67
12  55.0            89.0           1.65  32.69
14  56.0            76.0           1.69  26.61
17  56.0            74.0           1.69  25.91
21  56.0            76.0           1.69  26.61
23  56.0            76.0           1.69  26.61
7   57.0            34.0           1.61  13.12
22  67.0            78.0           1.85  22.79
15  67.0            78.0           1.85  22.79
18  67.0            78.0           1.85  22.79
4   68.0            50.0           1.32  28.70
20  68.0            50.0           1.32  28.70
8   75.0            23.0           1.24  14.96

In [107]:
df=df.reset_index(drop=True)

In [108]:
df

Age  Weight (in kg)  Height (in m)    BMI
0   10.0            32.0           1.65  11.75
1   12.0            43.0           1.21  29.37
2   21.0            43.0           1.52  18.61
3   23.0            45.0           1.75  14.69
4   23.0            53.0           1.50  23.56
5   26.0            65.0           1.21  44.40
6   26.0            65.0           1.21  44.40
7   26.0            65.0           1.21  44.40
8   32.0            21.0           1.52   9.09
9   34.0            65.0           1.76  20.98
10  45.0            60.0           1.35  32.92
11  54.0            78.0           1.50  34.67
12  55.0            89.0           1.65  32.69
13  56.0            76.0           1.69  26.61
14  56.0            74.0           1.69  25.91
15  56.0            76.0           1.69  26.61
16  56.0            76.0           1.69  26.61
17  57.0            34.0           1.61  13.12
18  67.0            78.0           1.85  22.79
19  67.0            78.0           1.85  22.79
20  67.0            78.0           1.85  22.79
21  68.0            50.0           1.32  28.70
22  68.0            50.0           1.32  28.70
23  75.0            23.0           1.24  14.96

In [115]:
# create a copy of the dataframe
new_df=df.copy()
new_df.head()

Age  Weight (in kg)  Height (in m)    BMI
0  10.0            32.0           1.65  11.75
1  12.0            43.0           1.21  29.37
2  21.0            43.0           1.52  18.61
3  23.0            45.0           1.75  14.69
4  23.0            53.0           1.50  23.56

In [116]:
# drop the BMI column
new_df.drop('BMI', axis=1, inplace=True)

In [118]:
# dropping the 23rd row
new_df.drop(23)

Age  Weight (in kg)  Height (in m)
0   10.0            32.0           1.65
1   12.0            43.0           1.21
2   21.0            43.0           1.52
3   23.0            45.0           1.75
4   23.0            53.0           1.50
5   26.0            65.0           1.21
6   26.0            65.0           1.21
7   26.0            65.0           1.21
8   32.0            21.0           1.52
9   34.0            65.0           1.76
10  45.0            60.0           1.35
11  54.0            78.0           1.50
12  55.0            89.0           1.65
13  56.0            76.0           1.69
14  56.0            74.0           1.69
15  56.0            76.0           1.69
16  56.0            76.0           1.69
17  57.0            34.0           1.61
18  67.0            78.0           1.85
19  67.0            78.0           1.85
20  67.0            78.0           1.85
21  68.0            50.0           1.32
22  68.0            50.0           1.32

## <span style="color:darkgreen;">Droping duplicates</span>

In [125]:
# Check if data has duplicates
duplicate_count =new_df.duplicated().sum()
duplicate_count

7

In [126]:
# drop duplicates from your data
new_df.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6      True
7      True
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15     True
16     True
17    False
18    False
19     True
20     True
21    False
22     True
23    False
dtype: bool

In [129]:
new_df.drop_duplicates(inplace=True)
new_df

Age  Weight (in kg)  Height (in m)
0   10.0            32.0           1.65
1   12.0            43.0           1.21
2   21.0            43.0           1.52
3   23.0            45.0           1.75
4   23.0            53.0           1.50
5   26.0            65.0           1.21
8   32.0            21.0           1.52
9   34.0            65.0           1.76
10  45.0            60.0           1.35
11  54.0            78.0           1.50
12  55.0            89.0           1.65
13  56.0            76.0           1.69
14  56.0            74.0           1.69
17  57.0            34.0           1.61
18  67.0            78.0           1.85
21  68.0            50.0           1.32
23  75.0            23.0           1.24

## <span style="color:darkgreen;">Checking for missing values</span>

Let's import a new dataset.

In [130]:
# Import missingdata.csv 
mdata = pd.read_csv('missingdata.csv')
mdata.head(2)

Age  Weight (in kg)  Height (in m)
0  45.0            60.0           1.35
1  12.0            43.0           1.21

In [131]:
# check for nulls
mdata.info()
mdata.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             22 non-null     float64
 1   Weight (in kg)  21 non-null     float64
 2   Height (in m)   22 non-null     float64
dtypes: float64(3)
memory usage: 680.0 bytes


Age               1
Weight (in kg)    2
Height (in m)     1
dtype: int64

The function `.isnull` check whether the data is missing. The `sum()` sums the number of 'True' values in the column. The final output gives the number of missing values in each column.

Here, we see there are 2 missing values in the 'weight' column and one missing value in other columns.

## <span style="color:crimson;">Take home exercise</span>

<a id="reading_data"> </a>
### Reading Data from Different Sources

Note that the files names are used as examples only. You can try importing your own files to execute the below examples.

**1. Read a `.xlsx` file**

`pd.read_excel('example.xlsx')`

**2. Read a `.txt` file**

`data = pd.read_csv('example.txt', sep="\t")`

**3. Read a `.zip` file**

`import zipfile
with zipfile.ZipFile('data.zip') as z:
    with z.open('example.csv') as f:
        file = pd.read_csv(f)
        print(file.head())`

**4. Read a `.html` file**

`df = pd.read_html('example.html', header=1, index_col=0)`

**5. Read a `.json` file**

`pd.read_json('example.json')`